lightning/src/chain/keysinterface.rs

```diff
@@ -469,56 +480,60 @@ impl ChannelKeys for InMemoryChannelKeys {
 	fn pubkeys(&self) -> &ChannelPublicKeys { &self.holder_channel_pubkeys }
 	fn key_derivation_params(&self) -> (u64, u64) { self.key_derivation_params }
 
-	fn sign_counterparty_commitment<T: secp256k1::Signing + secp256k1::Verification>(&self, feerate_per_kw: u32, commitment_tx: &Transaction, pre_keys: &PreCalculatedTxCreationKeys, htlcs: &[&HTLCOutputInCommitment], secp_ctx: &Secp256k1<T>) -> Result<(Signature, Vec<Signature>), ()> {
-		if commitment_tx.input.len() != 1 { return Err(()); }
-		let keys = pre_keys.trust_key_derivation();
+	fn sign_counterparty_commitment<T: secp256k1::Signing + secp256k1::Verification>(&self, commitment_tx: &CommitmentTransaction, secp_ctx: &Secp256k1<T>) -> Result<(Signature, Vec<Signature>), ()> {
+		let keys = commitment_tx.trust_key_derivation();
 
 		let funding_pubkey = PublicKey::from_secret_key(secp_ctx, &self.funding_key);
-		let accepted_data = self.accepted_channel_data.as_ref().expect("must accept before signing");
-		let channel_funding_redeemscript = make_funding_redeemscript(&funding_pubkey, &accepted_data.counterparty_channel_pubkeys.funding_pubkey);
-
-		let commitment_sighash = hash_to_message!(&bip143::SigHashCache::new(commitment_tx).signature_hash(0, &channel_funding_redeemscript, self.channel_value_satoshis, SigHashType::All)[..]);
-		let commitment_sig = secp_ctx.sign(&commitment_sighash, &self.funding_key);
-
-		let commitment_txid = commitment_tx.txid();
-
-		let mut htlc_sigs = Vec::with_capacity(htlcs.len());
-		for ref htlc in htlcs {
-			if let Some(_) = htlc.transaction_output_index {
-				let htlc_tx = chan_utils::build_htlc_transaction(&commitment_txid, feerate_per_kw, accepted_data.holder_selected_contest_delay, htlc, &keys.broadcaster_delayed_payment_key, &keys.revocation_key);
-				let htlc_redeemscript = chan_utils::get_htlc_redeemscript(&htlc, &keys);
-				let htlc_sighash = hash_to_message!(&bip143::SigHashCache::new(&htlc_tx).signature_hash(0, &htlc_redeemscript, htlc.amount_msat / 1000, SigHashType::All)[..]);
-				let our_htlc_key = match chan_utils::derive_private_key(&secp_ctx, &keys.per_commitment_point, &self.htlc_base_key) {
-					Ok(s) => s,
-					Err(_) => return Err(()),
-				};
-				htlc_sigs.push(secp_ctx.sign(&htlc_sighash, &our_htlc_key));
-			}
+		let channel_funding_redeemscript = make_funding_redeemscript(&funding_pubkey, &self.counterparty_pubkeys().funding_pubkey);
+
+		let channel_parameters = self.make_channel_parameters();
+		let directed_channel_parameters = channel_parameters.to_directed(false);
+		let commitment_sig = commitment_tx.get_signature(&directed_channel_parameters, &self.funding_key, &channel_funding_redeemscript, self.channel_value_satoshis, secp_ctx);
```

Comment:
This ends up creating the Tx, hashing it, and then dropping it, and then we create it again on the next line just to hash it, all in a call from `channel.rs` that is a few lines below `calculate_txid` in `send_commitment_no_state_update`. Given this is an incredibly common operation, I don't think that much malloc/free traffic (ie memory fragmentation potential) and SHA256 is acceptable. I'm still a bit dubious of the API here, its just too hard to accidentally blow up complexity for no reason - especially gnarly when it happens on the signer, potentially on embedded hardware.

If we want to change it, there are a few options - we could make the fields private and expose getters (eh? but we do already do this for the TxCreationKeys), or we could add a new struct that you can get from the `CommitmentTransaction` that holds a cached/calculated tx and then expose all the various signature/Txid methods on it instead. I think I would prefer the second, except its pretty hard to use that across the Channel/signer boundary, though there may be something we can do that's somewhere in between more akin to what we do for `TxCreationKeys`.

Reply:
Added a commit named "WIP: pre-build CommitmentTransaction" currently at https://github.com/rust-bitcoin/rust-lightning/pull/742/commits/b676b0cf97e3936312690fef34ce98c49db3f9ac.

I reached the conclusion that we might as well pre-build and save the bitcoin transaction, since we have to build the outputs for sorting in the constructor, and that's the O(n) work anyway - the rest is O(1).

Given that, the signer can trust the pre-built bitcoin tx, and ignore it if not by using the `build` function.

I moved the `get_signature` and `get_sighash` functions to the new `BuiltCommitmentTransaction` sub-structure.

I believe this eliminates duplicate tx builds and txid hashing.  Let me know how you feel about it.